core/templates/domain/topic_viewer/read-only-topic-object.factory.ts

```diff
@@ -0,0 +1,154 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating instances of ReadOnlyTopic from
+ * topic data domain objects.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { SkillSummary, SkillSummaryObjectFactory } from
+  'domain/skill/SkillSummaryObjectFactory';
+import { StoryReference, StoryReferenceObjectFactory } from
+  'domain/topic/StoryReferenceObjectFactory';
+import { StorySummary } from
+  'domain/story/StorySummaryObjectFactory';
+import { Subtopic, SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+
+export interface ISkillDescriptions {
+  [skillId: string]: string | null;
+}
+
+export interface IDegreesOfMastery {
+  [skillId: string]: number | null;
+}
+
+export class ReadOnlyTopic {
+  _topicName: string;
+  _topicId: string;
+  _canonicalStorySummaries: Array<StorySummary>;
+  _additionalStorySummaries: Array<StorySummary>;
+  _uncategorizedSkillSummaries: Array<SkillSummary>;
+  _subtopics: Array<Subtopic>;
+  _degreesOfMastery: IDegreesOfMastery;
+  _skillDescriptions: ISkillDescriptions;
+  _trainTabShouldBeDisplayed: boolean;
+
+  constructor(topicName: string, topicId: string,
+      canonicalStorySummaries: Array<StorySummary>,
+      additionalStorySummaries: Array<StorySummary>,
+      uncategorizedSkillSummaries: Array<SkillSummary>,
+      subtopics: Array<Subtopic>,
+      degreesOfMastery: IDegreesOfMastery,
+      skillDescriptions: ISkillDescriptions,
+      trainTabShouldBeDisplayed: boolean) {
+    this._topicName = topicName;
+    this._topicId = topicId;
+    this._canonicalStorySummaries = canonicalStorySummaries;
+    this._additionalStorySummaries = additionalStorySummaries;
+    this._uncategorizedSkillSummaries = uncategorizedSkillSummaries;
+    this._subtopics = subtopics;
+    this._degreesOfMastery = degreesOfMastery;
+    this._skillDescriptions = skillDescriptions;
+    this._trainTabShouldBeDisplayed = trainTabShouldBeDisplayed;
+  }
+
+  getTopicName(): string {
+    return this._topicName;
+  }
+
+  getTopicId(): string {
+    return this._topicId;
+  }
+
+  getCanonicalStorySummaries(): Array<StorySummary> {
+    return this._canonicalStorySummaries.slice();
+  }
+
+  getAdditionalStorySummaries(): Array<StorySummary> {
+    return this._additionalStorySummaries.slice();
+  }
+
+  getUncategorizedSkillsSummaries(): Array<SkillSummary> {
+    return this._uncategorizedSkillSummaries.slice();
+  }
+
+  getSubtopics(): Array<Subtopic> {
+    return this._subtopics.slice();
+  }
+
+  getDegreesOfMastery(): IDegreesOfMastery {
+    return this._degreesOfMastery;
```

Comment:
Do we need to return a coy of this dict as well? (not sure, can you please check?)

Reply:
Yes, its needed when you include skills.